decompiler/frontend/binaryninja/handlers/globals.py

```diff
@@ -47,133 +84,207 @@ def __init__(self, lifter):
             FloatType: self._lift_basic_type,
             BoolType: self._lift_basic_type,
             VoidType: self._lift_void_type,
-            ArrayType: self._lift_constant_type,
+            ArrayType: self._lift_array_type,
             PointerType: self._lift_pointer_type,
-            NamedTypeReferenceType: self._lift_named_type_ref,  # Lift DataVariable with type NamedTypeRef
+            NamedTypeReferenceType: self._lift_named_type_ref,
         }
+        self._lifted_globals: dict[int, GlobalVariable] = {}
+        self._view: Optional[BinaryView] = None
+        self._callers: list[int] = []
```

Comment:
`_lifted_globals` will have entries when the chain of global variables is already lifted (e.g. it goes from the most recent recursive call to the oldest one). => Can't be used to detect circles. Only needed when the same circle will be lifted twice (or more).

`_callers` will have entries when we lift the chain. E.g. When we lift A, we see that the adress of A is already part of the circle and stop there. => Needed to detect the circle in the first place

And yes when we lift A a second time, it will have the same chain as before.
Why would that be a problem for generating the code at the backend? 
The backend will collect all variables but strip the ssa labels of them, therefore duplicates will not occur. 
Or am I missing your point?

Reply:
Good to know that the cache is just for performance reasons. This should be explicitly noted in a comment.

My Point was, that when we lift A, it's lifted as A -> B -> C ->Symbol(A). When we next lift C, we get C -> Symbol(A) and not C -> A -> B -> Symbol(C). Therefore the length of these chains depends on the lifting order. But I'm not sure if this will negatively impact the output.

Reply:
Technically correct and the down side from `_lifted_globals`.
If we wanted to remove the problem completely we would need to step away from these chains of GlobalVariables in the first place. 

E.g. only lift each initial_value as a reference and store all referenced GlobalVariables inside a datastructure.
Don't really know if it is worth it for the struggle, because literally no stage looks at the initial_value.
Maybe it's a good discussion point before we implement structs?